concent_api/core/models.py

```diff
@@ -73,3 +76,167 @@ class MessageAuth(Model):
     message              = OneToOneField(StoredMessage, related_name = 'auth')
     provider_public_key  = Base64Field(max_length=64)
     requestor_public_key = Base64Field(max_length=64)
+
+
+class ClientManager(Manager):
+
+    def get_or_create_full_clean(self, public_key: bytes):
+        """
+        Returns Model instance.
+        Does the same as get_or_create method, but also performs full_clean() on newly created instance.
+        """
+        try:
+            instance = self.get(public_key = base64.b64encode(public_key))
```

Comment:
How about using get_or_create, which is atomic on database level ? The code might look like this then:
```
instance = self.model(
    public_key_bytes = public_key
)
instance.full_clean()
instance, _ = self.get_or_create(public_key = base64.b64encode(public_key))
return instance
```

Reply:
`get_or_create()` isn't any more atomic than our function. Maybe "atomic" is not the right word for what I meant here. Django docs say that [`get_or_create()` is "is atomic assuming correct usage, correct database configuration, and correct behavior of the underlying database"](https://docs.djangoproject.com/en/2.0/ref/models/querysets/#get-or-create) by which they mean that if the key is unique and the database supports transactions you will never end up with two copies in of the object in the database. All `get_or_create()` does to achieve that is wrap the "get" and "insert" parts in a single transaction. Our function is safe from that standpoint too because we use `ATOMIC_REQUESTS` so all our views run inside a transaction.

What I meant is that the side effect of this approach is that if there are two requests trying to do "get or create" at the same time, one with them may fail and raise `IntegrityError`. This is fine in the sense that the data is safe but it's undesirable for the client because it gets HTTP 500 and has to retry. Doing it with `INSERT ... ON CONFLICT` at the database level allows you to do this without the risk of an `IntegrityError`.

Reply:
`IntegrityError` can be caught too. How about this ?

Reply:
Yes. That's a valid solution for cases where the object is only created and never deleted.

We could catch it, try to get the object and raise an exception if that fails (because it means that the `IntegrityError` must have been caused by someting else - maybe the model has additional `unique` constraints that are not satisfied).

Anyway, I just wanted to point out the problem but it's not something that's urgent. We can create a separate issue for it.